kbl2vec/src/main/java/com/foursoft/harness/kbl2vec/utils/DeterministicXmlIdGenerator.java

```diff
@@ -0,0 +1,86 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * =========================LICENSE_END==================================
+ */
+package com.foursoft.harness.kbl2vec.utils;
+
+import com.foursoft.harness.kbl2vec.core.TransformationContext;
+import com.foursoft.harness.navext.runtime.model.Identifiable;
+import com.foursoft.harness.navext.runtime.model.ModifiableIdentifiable;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class DeterministicXmlIdGenerator extends XmlIdGenerator {
+
+    private final TransformationContext context;
+    private final Map<Object, Object> invertedEntityMapping;
+    private final Set<String> generatedIds;
+
+    public DeterministicXmlIdGenerator(final TransformationContext context) {
+        super();
+        this.context = context;
+        this.invertedEntityMapping = invertEntityMapping();
+        this.generatedIds = new HashSet<>();
+    }
+
+    private Map<Object, Object> invertEntityMapping() {
+        return context.getEntityMapping().getContent().entries().stream()
+                .collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey));
+    }
+
+    private String formatId(final String baseId, final int suffix) {
+        return baseId + "_" + suffix;
+    }
+
+    private String createIdWithCounter(final String baseId, int suffix) {
+        final String candidateId = formatId(baseId, suffix);
+        if (this.generatedIds.contains(candidateId)) {
+            return createIdWithCounter(baseId, ++suffix);
+        }
+        return candidateId;
+    }
+
+    @Override
+    protected void createIdForXmlBean(final ModifiableIdentifiable aBean, final String prefix) {
+        final Object source = invertedEntityMapping.get(aBean);
+        if (!(source instanceof final Identifiable identifiable)) {
+            super.createIdForXmlBean(aBean, prefix);
+            return;
+        }
+
+        final String vecXmlId = prefix + identifiable.getXmlId();
+        if (!this.generatedIds.contains(vecXmlId)) {
+            aBean.setXmlId(vecXmlId);
+            this.generatedIds.add(vecXmlId);
+            return;
+        }
```

Comment:
This code is unnecessary. It is all covered by the recursive call. It will behave the same without these lines. You will get Ids: `base_0`, `base_1`, etc. That was the reason why my suggestion had a condition for dropping the suffix when `suffix == 0` simulating the behaviour the first generated id is without suffix and then the first "duplicate" gets a "_1". I know, recursion is elegant, but sometimes hard to get in the brain. However, if you do the same (here the `contains` check) inside and outside the recursion, this is normally an indicator for an uncomplete recursion. 😉